MaterialLib/SolidModels/SolidsBase.h

```diff
@@ -0,0 +1,87 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef MATERIALLIB_SOLIDMODELS_SOLIDBASE_H_
+#define MATERIALLIB_SOLIDMODELS_SOLIDBASE_H_
+
+#include <memory>
+
+#include "ProcessLib/Deformation/BMatrixPolicy.h"
+#include "ProcessLib/Parameter/Parameter.h"
+
+namespace MeshLib
+{
+class Element;
+}
+
+namespace Solids
+{
+template <int DisplacementDim>
+struct SolidsBase
+{
+    struct InternalVariables
+    {
+        virtual void pushBackState() = 0;
+    };
+
+    using KelvinVector = ProcessLib::KelvinVectorType<DisplacementDim>;
+    using KelvinMatrix = ProcessLib::KelvinMatrixType<DisplacementDim>;
+
+    // TODO Avoid copies of data:
+    // Using MatrixBase<Derived> not possible because template functions cannot
+    // be virtual. Maybe there is a workaround for this.
+    // Using Map<Matrix<double, ...>> makes the interface (for the material
+    // model implementation) unnecessary difficult.
+    void computeConstitutiveRelation(
+        double const t,
+        ProcessLib::SpatialPosition const& x,
+        double const dt,
+        Eigen::Matrix<double, -1, 1> const& eps_prev,
+        Eigen::Matrix<double, -1, 1> const& eps,
+        Eigen::Matrix<double, -1, 1> const& sigma_prev,
+        Eigen::Matrix<double, -1, 1>& sigma,
```

Comment:
We cannot pull the computation of sigma and C apart. In the non-linear Burgers model, for example, sigma and C are the result of the same algorithm.
Also in the local assmebler part the residual and Jacobian are evaluated together, especially after the Newton-Raphson PR #1352.

In principle it is possible to separate those computations, but additional storage will be required to store the result of the local Newton iterations.